Add unique identifer to bmm thread_mm functions

dmpots opened the pull request:
Summary:
The bmm template generates code like this

```
template<bool accum>
void cpp_fused_bmm_66_micro_gemm(...) {
    ...
}

void single_thread_mm() {
    ...
    cpp_fused_bmm_66_micro_gemm(...)
    ...
}

void threaded_mm() {
    ...
    cpp_fused_bmm_66_micro_gemm(...)
    ...
}

void cpp_fused_bmm_66(...)
{
    ...
    single_thread_mm(...);
    ...
    threaded_mm(...);
    ...
}
```

The generated  `fused_bmm` and `fused_bmm_microgemm` functions both have unique identifiers added to their names, but the `single_threaded_mm` and `threaded_mm` do not.

This diff adds unique identifies to those generated functions as well. The identifier is based on the kernel name. So for the example above we would generate a bmm template name like `cpp_fused_bmm_66_single_thread_mm()`.

Differential Revision: D68364772


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

dmpots commented:
@jgong5 could you take a look at this PR? I'm not sure what's the best way to add a unit test for this. I ran into this on an internal workload.

frost-intel commented on torch/_inductor/codegen/cpp_bmm_template.py:
```diff
@@ -195,6 +195,7 @@ def get_options(
             if isinstance(sym, sympy.Expr)
             for s in sym.free_symbols
         ]
+        options["kernel_suffix"] = self.next_kernel_suffix()
```
Gemm code is already using `kernel.kernel_name` as a unique identifier. For consistency, could we use that instead? That way all the code pertaining to one bmm op will have a common identifier.

leslie-fang-intel replied:
yean, using `kernel.kernel_name` as prefix should align with existing implementation.

frost-intel replied:
No, we don't currently have any cpp wrapper tests for BMM. I added an example of one below that could be used to test this specific issue, but I'll work on adding any others that might be necessary.

leslie-fang-intel replied:
I think you can add the existing bmm template test into https://github.com/pytorch/pytorch/blob/d95a6babcc581ff06d1b914ee9f92c81b2e850e2/test/inductor/test_cpu_cpp_wrapper.py#L207-L216 enabled its testing with cpp wrapper.

leslie-fang-intel reviewed (commented):
Thanks for fixing this issue, could you also kindly adding a test case in `test_cpu_select_algorithm.py`?

frost-intel commented:
Here's a unit test that could be added in `test/inductor/test_cpu_select_algorithm.py` within `TestSelectAlgorithm`:

```python
    @patches
    @torch.no_grad
    @unittest.skipIf(not TEST_MKL, "Test requires MKL")
    @dtypes(torch.float)
    def test_aoti_bmm_unique_identifiers(self, dtype):
        try:
            try:
                from . import test_aot_inductor_utils
            except ImportError:
                import test_aot_inductor_utils
        except Exception:
            # skip this UT if import failed
            return

        class M(torch.nn.Module):
            def __init__(self):
                super().__init__()

            def forward(self, x, w):
                y = x @ w
                return y @ w

        counters.clear()
        x = torch.randn(3, 64, 64).to(dtype=dtype)
        w = torch.randn(3, 64, 64).to(dtype=dtype)
        mod = M().to(dtype=dtype).eval()
        with verify(dtype) as (atol, rtol), torch.no_grad():
            expected = mod(x, w)
            actual = test_aot_inductor_utils.AOTIRunnerUtil.run(
                "cpu",
                mod,
                (x, w),
            )
            self.assertEqual(actual, expected, atol=atol, rtol=rtol)
        self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 2)
```

dmpots commented:
> Here's a unit test that could be added in `test/inductor/test_cpu_select_algorithm.py` within `TestSelectAlgorithm`:
@frost-intel  thanks so much for the help with a test! I confirmed this fails without the change and added it to the PR.

dmpots commented:
> Thanks for fixing this issue, could you also kindly adding a test case in `test_cpu_select_algorithm.py`?


@leslie-fang-intel I added a unit test with help from @frost-intel.

leslie-fang-intel reviewed (approved):
Thanks for the fix. LGTM.

dmpots commented:
@pytorchbot label "topic: not user facing"

dmpots committed:
Add unique identifer to bmm thread_mm functions (#145303)

Summary:

The bmm template generates code like this

```
template<bool accum>
void cpp_fused_bmm_66_micro_gemm(...) {
    ...
}

void single_thread_mm() {
    ...
    cpp_fused_bmm_66_micro_gemm(...)
    ...
}

void threaded_mm() {
    ...
    cpp_fused_bmm_66_micro_gemm(...)
    ...
}

void cpp_fused_bmm_66(...)
{
    ...
    single_thread_mm(...);
    ...
    threaded_mm(...);
    ...
}
```

The generated  `fused_bmm` and `fused_bmm_microgemm` functions both have unique identifiers added to their names, but the `single_threaded_mm` and `threaded_mm` do not.

This diff adds unique identifies to those generated functions as well. The identifier is based on the kernel name. So for the example above we would generate a bmm template name like `cpp_fused_bmm_66_single_thread_mm()`.

Differential Revision: D68364772

hl475 commented:
@pytorchbot merge
